typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-list.rkt

```diff
@@ -78,21 +78,29 @@
         ;; TODO fix double typechecking
         [_ (tc/app-regular #'form expected)])))
   ;; special case for `list'
-  (pattern (list . args)
-    (let ()
+  (pattern
```

Comment:
I think this code could have some much simpler special cases based on the expected type (for example, if it's `(List A B C)` just use those as the expected type for each list element, or if it's `(U (List A B C) (List D E F))` then try inference with each union element.